### ISSUE TITLE
Test-suite sometimes fails with an odd error

### ISSUE THREAD

**Issue by AndrejMitrovic** [bosagora/agora#864]
```
Run ./ci/run.sh
+ set -o pipefail
+ dub test -b unittest-cov --skip-registry=all --compiler=ldc2
Executable configuration "unittest" of package agora defines no main source file, this may cause certain build modes to fail. Add an explicit "mainSourceFile" to the package description to fix this.
Running custom 'unittest' configuration.
Running pre-generate commands for vibe-d:tls...
Running pre-generate commands for agora...
Failed to execute program (No such file or directory)
Command failed with exit code 2: /Users/runner/hostedtoolcache/dc/ldc2-4eaa2fd8/x64/ldc2-4eaa2fd8-osx-x86_64/bin/dub source/scpp/build.d
```

This is a strange one, it seems like there's a race condition somewhere where we try to run an executable which doesn't exist yet. I see it happen more and more often.

**Comment by AndrejMitrovic** [comment]
https://github.com/bpfkorea/agora/pull/845/checks?check_run_id=695047246

**Comment by TrustHenry** [comment]
If that's the case, If you print out the command of build.d, it will become clearer.
https://github.com/bpfkorea/agora/blob/c3eef128192bff125a868c7d6be7c16d97bdd5b1/source/scpp/build.d#L185

**Comment by Geod24** [comment]
There's only three possibilities I can think of:
- `cpp` doesn't exists;
- `dub` doesn't exists;
- `source/scpp/build.d` doesn't exists;

Now one of the thing you would see, if the command failed, was [this](https://github.com/bpfkorea/agora/blob/c3eef128192bff125a868c7d6be7c16d97bdd5b1/source/scpp/build.d#L153-L154). Since we don't see it, uncommenting `writeln` won't do anything because the failure is before that.

**Comment by Geod24** [comment]
I am starting to suspect cache pollution in the Github action.

**Comment by AndrejMitrovic** [comment]
So the issue is multiple jobs running in parallel?

**Comment by Geod24** [comment]
I will test this theory shortly

**Comment by Geod24** [comment]
Just got this:
```
+ dub build --skip-registry=all --compiler=ldc2
Running pre-generate commands for vibe-d:tls...
Running pre-generate commands for agora...
Failed to execute '../../../../../var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/.dub/build/version_build-~master/application-debug-posix.osx.darwin-x86_64-ldc_v1.25.0-D8B3A9F66D677C52DEFEC21E4B5261F2/version_build' (No such file or directory)
Command failed with exit code 2: /Users/runner/hostedtoolcache/dc/ldc2-1.25.0/x64/ldc2-1.25.0-osx-x86_64/bin/dub source/agora/cli/version/main.d
```

Seems like a race condition between compilation finishing and running the process ?

**Comment by Geod24** [comment]
```
+ ldc2 -i -run ./tests/runner.d ldc2 -cov
[fatal]	Failed to execute '../../../../../var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/.dub/build/tracy_build-~master/application-debug-posix.osx.darwin-x86_64-ldc_v1.25.0-6CA22FC09134B9803A88157F6581C057/tracy_build' (No such file or directory)
Error: /var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/runner-5fd42e failed with status: 1
[fatal]	Target 'agora' doesn't exist. Is the target type set to "none" in the package recipe?
Error: Process completed with exit code 1.
```

Those new outputs might shed some light on the issue.

**Comment by Geod24** [comment]
```
Run ./ci/run.sh
+ set -o pipefail
+ export AGORA_VERSION=HEAD
+ AGORA_VERSION=HEAD
+ dub build --skip-registry=all --compiler=ldc2
Running pre-generate commands for vibe-d:tls...
Failed to execute '../../../../../../../../var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/.dub/build/script-~master/application-debug-posix.osx.darwin-x86_64-ldc_v1.25.0-CFEAF3292F597EE78CBF890152994412/script' (No such file or directory)
Running pre-generate commands for agora...
Failed to execute '../../../../../var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/.dub/build/cpp_build-~master/application-debug-posix.osx.darwin-x86_64-ldc_v1.25.0-8FED1FEB53BE1FF674218C8646405698/cpp_build' (No such file or directory)
Failed to execute '../../../../../var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/.dub/build/cpp_build-~master/application-debug-posix.osx.darwin-x86_64-ldc_v1.25.0-8FED1FEB53BE1FF674218C8646405698/cpp_build' (No such file or directory)
Command failed with exit code 2: /Users/runner/hostedtoolcache/dc/ldc2-1.25.0/x64/ldc2-1.25.0-osx-x86_64/bin/dub source/scpp/build.d || (sleep 5s && /Users/runner/hostedtoolcache/dc/ldc2-1.25.0/x64/ldc2-1.25.0-osx-x86_64/bin/dub source/scpp/build.d)
Error: Process completed with exit code 2.
```

So this seems to be something that has to do with `preGenerateCommands` (and maybe other `pre` hooks, but we don't use them). We currently have a retry in place to mitigate the issue, but it *still* shows up:
https://github.com/bosagora/agora/blob/903ff9d895554f0dc48d544f78f6dd761b46eb1a/dub.json#L13-L16

**Comment by omerfirmak** [comment]
To me it seems like dub is failing to create some temp directory it needs.
```
Failed to execute '../../../../../var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/.dub/build/cpp_build-~master/application-debug-posix.osx.darwin-x86_64-ldc_v1.25.0-8FED1FEB53BE1FF674218C8646405698/cpp_build' (No such file or directory)
```

**Comment by Geod24** [comment]
Worked around in #1874 and should not be a problem, closing.

**Comment by Geod24** [comment]
```
+ ldc2 -i -run ./tests/runner.d ldc2 -cov
[fatal]	Using dub registry url 'https://code.dlang.org/'
Error: /var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/runner-85ed1a failed with status: 1
[fatal]	Refreshing local packages (refresh existing: true)...
[fatal]	Looking for local package map at /var/lib/dub/packages/local-packages.json
[fatal]	Looking for local package map at /Users/runner/.dub/packages/local-packages.json
[fatal]	Looking for local package map at /Users/runner/work/agora/agora/.dub/packages/local-packages.json
[fatal]	Note: Failed to determine version of package agora at .. Assuming ~master.
[fatal]	Refreshing local packages (refresh existing: false)...
[fatal]	Looking for local package map at /var/lib/dub/packages/local-packages.json
[fatal]	Looking for local package map at /Users/runner/.dub/packages/local-packages.json
[fatal]	Looking for local package map at /Users/runner/work/agora/agora/.dub/packages/local-packages.json
[fatal]	Refreshing local packages (refresh existing: false)...
[fatal]	Looking for local package map at /var/lib/dub/packages/local-packages.json
[fatal]	Looking for local package map at /Users/runner/.dub/packages/local-packages.json
[fatal]	Looking for local package map at /Users/runner/work/agora/agora/.dub/packages/local-packages.json
[fatal]	Generating using build
[fatal]	Configuring dependent version_build, deps:
[fatal]	Using existing build in /Users/runner/work/agora/agora/source/agora/cli/version/.dub/build/application-debug-posix.osx.darwin-x86_64-ldc_v1.25.0-D8B3A9F66D677C52DEFEC21E4B5261F2/.
[fatal]	Copying target from /Users/runner/work/agora/agora/source/agora/cli/version/.dub/build/application-debug-posix.osx.darwin-x86_64-ldc_v1.25.0-D8B3A9F66D677C52DEFEC21E4B5261F2/version_build to /Users/runner/work/agora/agora/source/agora/cli/version
[fatal]	Using dub registry url 'https://code.dlang.org/'
[fatal]	Refreshing local packages (refresh existing: true)...
[fatal]	Looking for local package map at /var/lib/dub/packages/local-packages.json
[fatal]	Looking for local package map at /Users/runner/.dub/packages/local-packages.json
[fatal]	Looking for local package map at /Users/runner/work/agora/agora/.dub/packages/local-packages.json
[fatal]	Note: Failed to determine version of package agora at .. Assuming ~master.
[fatal]	Refreshing local packages (refresh existing: false)...
[fatal]	Looking for local package map at /var/lib/dub/packages/local-packages.json
[fatal]	Looking for local package map at /Users/runner/.dub/packages/local-packages.json
[fatal]	Looking for local package map at /Users/runner/work/agora/agora/.dub/packages/local-packages.json
[fatal]	Refreshing local packages (refresh existing: false)...
[fatal]	Looking for local package map at /var/lib/dub/packages/local-packages.json
[fatal]	Looking for local package map at /Users/runner/.dub/packages/local-packages.json
[fatal]	Looking for local package map at /Users/runner/work/agora/agora/.dub/packages/local-packages.json
[fatal]	Generating using build
[fatal]	Configuring dependent cpp_build, deps:
[fatal]	Using existing build in /Users/runner/work/agora/agora/source/scpp/.dub/build/application-debug-posix.osx.darwin-x86_64-ldc_v1.25.0-8FED1FEB53BE1FF674218C8646405698/.
[fatal]	Copying target from /Users/runner/work/agora/agora/source/scpp/.dub/build/application-debug-posix.osx.darwin-x86_64-ldc_v1.25.0-8FED1FEB53BE1FF674218C8646405698/cpp_build to /Users/runner/work/agora/agora/source/scpp
[fatal]	Failed to execute '../../../../../var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/.dub/build/tracy_build-~master/application-debug-posix.osx.darwin-x86_64-ldc_v1.25.0-6CA22FC09134B9803A88157F6581C057/tracy_build' (No such file or directory)
[fatal]	Target 'agora' doesn't exist. Is the target type set to "none" in the package recipe?
Error: Process completed with exit code 1.
```

**Comment by Geod24** [comment]
So, turns out that Github does some buffering on the output that is sent to stdout, which results in some output being lost when the program terminates (just hit it with LocalRest).
I don't know by which mechanism it happens yet (reported as https://github.com/actions/runner/issues/1235) but it could be the source of the spurious failures if we take [this](https://github.com/dlang/dub/pull/1998#issuecomment-694634112) into consideration.
